icubam/backoffice/server_test.py

```diff
@@ -119,3 +120,95 @@ def test_upload(self):
     self.assertIn('error', resp_data)
     self.assertIn('msg', resp_data)
     self.assertEqual(resp_data['error'], False)
+
+  def test_post_region(self):
+    handler = regions.RegionHandler
+    with mock.patch.object(base.BaseHandler, 'get_current_user') as m:
+      m.return_value = self.admin
+
+      name = "InvalidRegion"
+      self.assertIsNone(self.db.get_region_by_name(name))
+      response = self.fetch(
+        handler.ROUTE, method='POST', body=f"region_id=&name={name}"
+      )
+      # redirect to ListRegionsHandler
+      self.assertEqual(response.code, 302)
+      self.assertIsNotNone(self.db.get_region_by_name(name))
+
+  def test_post_icu(self):
+    handler = icus.ICUHandler
+    with mock.patch.object(base.BaseHandler, 'get_current_user') as m:
+      m.return_value = self.admin
+
+      data = {
+        'icu_id': '',
+        'name': 'TestICU',
+        'telephone': '00000000000',
+        'legal_id': '',
+        'is_active': 'on',
+        'region_id': 1,
+        'city': 'Paris',
+        'dept': 'Ile-de-France',
+        'lat': 0.0000,
+        'long': 0.0000
+      }
+      self.assertIsNone(self.db.get_icu_by_name(data['name']))
+      response = self.fetch(
+        handler.ROUTE,
+        method='POST',
+        body='&'.join(f'{key}={val}' for key, val in data.items())
+      )
+      # redirect to ListICUsHandler
+      self.assertEqual(response.code, 302)
+      self.assertIsNotNone(self.db.get_icu_by_name(data['name']))
+
+  def test_post_token(self):
+    handler = tokens.TokenHandler
+    with mock.patch.object(base.BaseHandler, 'get_current_user') as m:
+      m.return_value = self.admin
+
+      data_email = 'test@test.org'
+      data = {
+        'external_client_id': '',
+        'name': 'TestToken',
+        'telephone': '00000000000',
+        'email': "test%40test.org",
+        'is_active': 'on',
+        'access_type': 'ALL',
+        'expiration_date': ''
+      }
+      self.assertIsNone(self.db.get_external_client_by_email(data_email))
+      response = self.fetch(
+        handler.ROUTE,
+        method='POST',
+        body='&'.join(f'{key}={val}' for key, val in data.items())
+      )
+      # redirect to ListTokensHandler
+      self.assertEqual(response.code, 302)
+      self.assertIsNotNone(self.db.get_external_client_by_email(data_email))
+
+  def test_post_user(self):
+    handler = users.UserHandler
+    with mock.patch.object(base.BaseHandler, 'get_current_user') as m:
+      m.return_value = self.admin
+
+      data_email = 'test@test.org'
+      data = {
+        'user_id': '',
+        'name': 'TestUser',
+        'telephone': '00000000000',
+        'is_active': 'on',
+        'email': "test%40test.org",
+        'password': 'test123',
+        'icus%5B%5D': '1',
```

Comment:
`%5B%5D` are `[` and `]` likely to indicate this value as a list. Not sure how this works exactly but in any case that's the value that gets send to the server during a POST request currently.

Reply:
Yes it is the way the post requests send a list and not a single value (the first one).
By the way, i would prefere to use `tornado.escape.url_escape('icus[]')` instead, or rather define your data dict, and then `{tornado.escupe.url_escape(k): tornado.escupe.url_escape(v) for (k, v) in data.items()}`

Reply:
or below when calling self.fetch

Reply:
Yes, I considered escaping as well. Will do. Thanks for the review!

Reply:
Actually, `urllib.parse.urlencode` can both convert dict to query and escape.